### PR TITLE
add support for write_lpex

### PR DESCRIPTION
**Pull request by tobikris**
This adds support for dumping the model in lp format to file and as a string.

### CODE REVIEW

File: golpa.go [hunk 1]
```diff
@@ -404,3 +407,45 @@ func (model *Model) SolveWithContext(ctx context.Context) (res *SolveResult, err
...
+	defer model.mu.RUnlock()
+
+	ret := C.write_lp(model.prob, C.CString(filename))
+
+	if ret != C.TRUE {
+		return fmt.Errorf("model not written successfully")
+	}
+
+	return nil
+}
```
Comment (costela) [inline golpa.go]: I don't think we need to increase the API surface with two "dump" methods. The second one (returning the content) is flexible enough and feels a lot more idiomatic, so I think we can get rid of this one.

File: golpa.go [hunk 2]
```diff
@@ -404,3 +407,45 @@ func (model *Model) SolveWithContext(ctx context.Context) (res *SolveResult, err
...
+
+	return 0
+}
+
+// WriteLP returns the model in lp format.
+func (model *Model) WriteLP() (string, error) {
+	model.mu.RLock()
+	defer model.mu.RUnlock()
+
+	model.writeBuf.Reset()
```
Comment (costela) [inline golpa.go]: do we really need to keep the buffer in the model? Can't we just instantiate it in the method and use `saveRef`/`loadRef` directly on it?

File: golpa.go [hunk 3]
```diff
@@ -404,3 +407,45 @@ func (model *Model) SolveWithContext(ctx context.Context) (res *SolveResult, err
...
+		return 0
+	}
+
+	model.writeBuf.WriteString(C.GoString(buf))
+
+	return 0
+}
+
+// WriteLP returns the model in lp format.
+func (model *Model) WriteLP() (string, error) {
```
Comment (costela) [inline golpa.go]: nitpicking: since we're not really writing anything, can't we call this `DumpLP` or `ExportLP`?

**Review by costela — CHANGES_REQUESTED** [review]
Thanks for the code! Just left a few thoughts. PTAL

File: golpa.go (hunk 3 above)
Comment (tobikris) [reply golpa.go]: I agree that this name is weird, but I wanted to stay as close as possible to the lpsolve naming. I think this is helping the user.

File: golpa.go (hunk 1 above)
Comment (tobikris) [reply golpa.go]: Ok, I removed this one.

File: golpa.go (hunk 2 above)
Comment (tobikris) [reply golpa.go]: Good point, changed it.

File: golpa.go (hunk 3 above)
Comment (tobikris) [reply golpa.go]: I guess you don't agree? Should I change it to `ExportLP`?

Comment (costela) [reply golpa.go]: oops, sorry, went under the radar. Yeah, the current API isn't really tied to lpsolve (switched from glpk to lpsolve in the past; might switch again in the future). So I'd rather have something the "reads" well, than mimic lpsolve.

Comment (tobikris) [reply golpa.go]: I see - changed it to ExportLP.

**Review by costela — APPROVED** [review]

**Comment by costela** [comment]
:rocket:

**Pull request merged**